src/concurrency/timestamp_ordering_transaction_manager.cpp

```diff
@@ -397,83 +184,286 @@ bool TimestampOrderingTransactionManager::AcquireOwnership(
-      // Can not own
-      return false;
-    }
-    if (AcquireOwnership(current_txn, tile_group_header, tuple_id) == false) {
-      // Can not acquire ownership
-      return false;
+  // TODO: what if we want to read a version that we write? 
+  else if (current_txn->GetIsolationLevel() == IsolationLevelType::SNAPSHOT) {
+  
+    oid_t tile_group_id = location.block;
+    oid_t tuple_id = location.offset;
+
+    LOG_TRACE("PerformRead (%u, %u)\n", location.block, location.offset);
+    auto &manager = catalog::Manager::GetInstance();
+    auto tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
+
+    // Check if it's select for update before we check the ownership 
+    // and modify the last reader cid
+    if (acquire_ownership == true) {
+
+      // get the latest version of this tuple.
+      location = *(tile_group_header->GetIndirection(location.offset));
+
+      tile_group_id = location.block;
+      tuple_id = location.offset;
+
+      tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
+
+      if (IsOwner(current_txn, tile_group_header, tuple_id) == false) {
+
+        // Acquire ownership if we haven't
+        if (IsOwnable(current_txn, tile_group_header, tuple_id) == false) {
+          // Cannot own
+          return false;
+        }
+        if (AcquireOwnership(current_txn, tile_group_header, tuple_id) == false) {
+          // Cannot acquire ownership
+          return false;
+        }
+
+        // Record RWType::READ_OWN
+        current_txn->RecordReadOwn(location);
+      }
+
+      // if we have already owned the version.
+      PL_ASSERT(IsOwner(current_txn, tile_group_header, tuple_id) == true);
+      
+      // Increment table read op stats
+      if (FLAGS_stats_mode != STATS_TYPE_INVALID) {
+        stats::BackendStatsContext::GetInstance()->IncrementTableReads(
+            location.block);
+      }
+
+      return true;
+    
+    } else {
+      // if it's not select for update, then update read set and return true.
+      
+      current_txn->RecordRead(location);
+
+      // Increment table read op stats
+      if (FLAGS_stats_mode != STATS_TYPE_INVALID) {
+        stats::BackendStatsContext::GetInstance()->IncrementTableReads(
+            location.block);
+      }
+      return true;
     }
-    // Promote to RWType::READ_OWN
-    current_txn->RecordReadOwn(location);
-  }
 
-  // if the current transaction has already owned this tuple, then perform read
-  // directly.
-  if (IsOwner(current_txn, tile_group_header, tuple_id) == true) {
-    PL_ASSERT(GetLastReaderCommitId(tile_group_header, tuple_id) <=
-              current_txn->GetBeginCommitId());
-    // Increment table read op stats
-    if (FLAGS_stats_mode != STATS_TYPE_INVALID) {
-      stats::BackendStatsContext::GetInstance()->IncrementTableReads(
-          location.block);
+  } // end SNAPSHOT
+
+  //////////////////////////////////////////////////////////
+  //// handle READ_COMMITTED
```

Comment:
I think we'd better keep them in separate so as to make the logic clean enough for review.